client/app/pages/data-sources/DataSourcesList.jsx

```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+import Button from 'antd/lib/button';
+import { react2angular } from 'react2angular';
+import { isEmpty, get } from 'lodash';
+import settingsMenu from '@/services/settingsMenu';
+import { DataSource, IMG_ROOT } from '@/services/data-source';
+import { policy } from '@/services/policy';
+import navigateTo from '@/services/navigateTo';
+import { $route } from '@/services/ng';
+import { routesToAngularRoutes } from '@/lib/utils';
+import CardsList from '@/components/cards-list/CardsList';
+import LoadingState from '@/components/items-list/components/LoadingState';
+import CreateSourceDialog from '@/components/CreateSourceDialog';
+import helper from '@/components/dynamic-form/dynamicFormHelper';
+
+class DataSourcesList extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { dataSourceTypes: [], dataSources: [], loading: true };
+  }
+
+  componentDidMount() {
+    Promise.all([
+      DataSource.query().$promise,
+      DataSource.types().$promise,
+    ]).then(values => this.setState({
+      dataSources: values[0],
+      dataSourceTypes: values[1],
+      loading: false,
+    }, () => { // all resources are loaded in state
+      if ($route.current.locals.isNewDataSourcePage) {
+        this.showCreateSourceDialog();
+      }
+    }));
+  }
+
+  createDataSource = (selectedType, values) => {
+    const target = { options: {}, type: selectedType };
+    helper.updateTargetWithValues(target, values);
+
+    return DataSource.save(target).$promise.then(() => {
+      this.setState({ loading: true });
+      DataSource.query(dataSources => this.setState({ dataSources, loading: false }));
+    }).catch((error) => {
+      if (!(error instanceof Error)) {
+        error = new Error(get(error, 'data.message', 'Failed saving.'));
+      }
+      return Promise.reject(error);
+    });
+  };
+
+  showCreateSourceDialog = () => {
+    CreateSourceDialog.showModal({
+      types: this.state.dataSourceTypes,
+      sourceType: 'Data Source',
+      imageFolder: IMG_ROOT,
+      helpTriggerPrefix: 'DS_',
+      onCreate: this.createDataSource,
+    }).result.then((success) => {
+      if (success) {
+        this.setState({ loading: true });
+        DataSource.query(dataSources => this.setState({ dataSources }));
+      }
+    }).finally(() => {
+      if ($route.current.locals.isNewDataSourcePage) {
+        navigateTo('data_sources');
+      }
+    });
```

Comment:
No need for state `{ loading: false }` ?

Reply:
I'm currently setting `loading: false` only for success cases. Error cases should be caught by angular, but I realized they are not :thinking:.